veniq/dataset_collection/dataflow/annotation.py

```diff
@@ -0,0 +1,411 @@
+    :return:
+    """
+    m_decl_names = set([x.name for x in method_declaration.parameters])
+    m_inv_names = set([x.member for x in invocaton_node.arguments])
+    intersection = m_inv_names.difference(m_decl_names)
+    if not intersection:
+        return True
+    else:
+        return False
+
+
+def determine_algorithm_insertion_type(
+        extracted_method_decl: ASTNode) -> InlineTypesAlgorithms:
+    has_attr_return_type = hasattr(extracted_method_decl, 'return_type')
+    if has_attr_return_type:
+        if not extracted_method_decl.return_type:
+            return InlineTypesAlgorithms.WITHOUT_RETURN_WITHOUT_ARGUMENTS
+        else:
+            return InlineTypesAlgorithms.WITH_RETURN_WITHOUT_ARGUMENTS
+    #  Else if we have constructor, it doesn't have return type
+    else:
+        return InlineTypesAlgorithms.WITHOUT_RETURN_WITHOUT_ARGUMENTS
+
+
+def annotate(dct: Dict[str, Any]):
+    method_invoked = dct['method_invoked']
+    target_node = dct['target_node']
+    extracted_m_decl = dct['extracted_m_decl']
+    ast = dct['ast']
+    text = dct['text']
+    if method_invoked.parent.node_type == ASTNodeType.THIS:
+        parent = method_invoked.parent.parent
+        class_names = [x for x in method_invoked.parent.children if hasattr(x, 'string')]
+        member_references = [x for x in method_invoked.parent.children if hasattr(x, 'member')]
+        lst = [x for x in member_references if x.member != method_invoked.member] + class_names
+        no_children = not lst
+    else:
+        parent = method_invoked.parent
+        no_children = True
+
+    is_not_is_extract_method_abstract = True
+    if 'abstract' in extracted_m_decl.modifiers:
+        is_not_is_extract_method_abstract = False
+
+    maybe_if = parent.parent
+    is_not_method_inv_single_statement_in_if = True
+    if maybe_if.node_type == ASTNodeType.IF_STATEMENT:
+        if hasattr(maybe_if.then_statement, 'expression'):
+            if maybe_if.then_statement.expression.node_type == ASTNodeType.METHOD_INVOCATION:
+                is_not_method_inv_single_statement_in_if = False
+
+    ast_subtree_method_decl = ast.get_subtree(extracted_m_decl)
+    # is_not_assign_value_with_return_type = True
+    is_not_several_returns = True
+    if hasattr(extracted_m_decl, 'return_type'):
+        if extracted_m_decl.return_type:
+            return_stats = len([
+                x for x in extracted_m_decl.body
+                if x.node_type == ASTNodeType.RETURN_STATEMENT]
+            )
+
+            # If we do not have return in function body,
+            # it means that we will have deep inside AST tree, so remember it
+            stats = [x for x in ast_subtree_method_decl.get_proxy_nodes(ASTNodeType.RETURN_STATEMENT)]
+            total_return_statements = len(stats) - return_stats
+            if total_return_statements > 0:
+                is_not_several_returns = False
+
+    has_not_throw = len([
```

Comment:
where throw is last element

func() {
    if () {
    }
    throw Exception();
}


func() {
    if () {
          throw Exception();
    }
}